File: mobile/android/app/src/main/java/app/serika/musicy/mobile/sync/SyncClient.kt
package app.serika.musicy.mobile.sync

import android.util.Log
import app.serika.musicy.mobile.data.MusicyRepository
import app.serika.musicy.mobile.data.api.ApiClient
import app.serika.musicy.mobile.data.model.DeviceListEvent
import app.serika.musicy.mobile.data.model.SyncClaimEvent
import app.serika.musicy.mobile.data.model.SyncCommandEvent
import app.serika.musicy.mobile.data.model.SyncCommandPayload
import app.serika.musicy.mobile.data.model.SyncDevice
import app.serika.musicy.mobile.data.model.SyncStateEvent
import app.serika.musicy.mobile.data.model.SyncStatePayload
import app.serika.musicy.mobile.data.model.Track
import kotlin.math.min
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Musicy Connect for Android — the phone joins the same device bus the web app
 * uses, so playback can be handed between a browser tab and this app.
 *
 * Transport is the server's SSE stream (`/api/sync/stream`) read straight off
 * an OkHttp response body; commands go back over `/api/sync/publish`.
 */
class SyncClient(
    private val repo: MusicyRepository,
    private val scope: CoroutineScope
) {
    private val json = ApiClient.json

    private val _connected = MutableStateFlow(false)
    val connected: StateFlow<Boolean> = _connected.asStateFlow()

    private val _devices = MutableStateFlow<List<SyncDevice>>(emptyList())
    val devices: StateFlow<List<SyncDevice>> = _devices.asStateFlow()

    private val _activeDeviceId = MutableStateFlow<String?>(null)
    val activeDeviceId: StateFlow<String?> = _activeDeviceId.asStateFlow()

    /** Last state broadcast by whichever device currently holds playback. */
    private val _remoteState = MutableStateFlow<SyncStatePayload?>(null)
    val remoteState: StateFlow<SyncStatePayload?> = _remoteState.asStateFlow()

    private val _deviceId = MutableStateFlow("")
    val deviceId: StateFlow<String> = _deviceId.asStateFlow()

    /** Invoked for commands addressed to this device. Runs off the main thread. */
    var onCommand: ((SyncCommandPayload) -> Unit)? = null

    /** Invoked when another device claims playback, so we can stop our audio. */
    var onRemoteClaim: ((String) -> Unit)? = null

    private var streamJob: Job? = null
    private var deviceName: String = "Android device"

    val isThisDeviceActive: Boolean
        get() = _activeDeviceId.value != null && _activeDeviceId.value == _deviceId.value

    fun start() {
        if (streamJob?.isActive == true) return
        streamJob = scope.launch(Dispatchers.IO) {
            repo.awaitConfig()
            val id = repo.settingsStore.ensureDeviceId()
            _deviceId.value = id
            deviceName = repo.settings.value.deviceName.ifBlank { "Android device" }

            var attempt = 0
            while (isActive) {
                if (!repo.settings.value.syncEnabled) {
                    _connected.value = false
                    delay(5_000)
                    continue
                }
                try {
                    openStream(id)
                    attempt = 0
                } catch (cancelled: kotlinx.coroutines.CancellationException) {
                    throw cancelled
                } catch (e: Exception) {
                    Log.w(TAG, "sync stream dropped: ${e.message}")
                } finally {
                    _connected.value = false
                }
                ensureActive()
                // Exponential backoff with jitter, capped, mirroring the web client.
                val backoff = min(15_000L, 1_000L * (1L shl min(attempt, 4)))
                attempt++
                delay(backoff + (0..500).random())
            }
        }
        // Seed the picker from the REST endpoint so it is populated before the
        // first `device-list` frame arrives.
        scope.launch(Dispatchers.IO) {
            runCatching { repo.devices() }.getOrNull()?.let { list ->
                _devices.value = list
                list.firstOrNull { it.isActive }?.let { _activeDeviceId.value = it.id }
            }
        }
    }

    fun stop() {
        streamJob?.cancel()
        streamJob = null
        _connected.value = false
    }

    private suspend fun openStream(id: String) = withContext(Dispatchers.IO) {
        val config = repo.config.value
        val base = ApiClient.normalizedBaseUrl(config)
        val url = "$base/api/sync/stream?deviceId=${enc(id)}&name=${enc(deviceName)}"
        val request = Request.Builder()
            .url(url)
            .header("Accept", "text/event-stream")
            .header("Cache-Control", "no-cache")
            .build()

        ApiClient.okHttp(config).newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("sync stream HTTP ${response.code}")
            _connected.value = true
            val source = response.body?.source() ?: error("empty sync stream")

            var event = "message"
            val data = StringBuilder()
            while (isActive && !source.exhausted()) {
                val line = source.readUtf8LineStrict()
                when {
                    line.isEmpty() -> {
                        if (data.isNotEmpty()) dispatch(event, data.toString())
                        event = "message"
                        data.setLength(0)
                    }
                    line.startsWith(":") -> Unit // heartbeat comment
                    line.startsWith("event:") -> event = line.removePrefix("event:").trim()
                    line.startsWith("data:") -> data.append(line.removePrefix("data:").trim())
                }
            }
        }
    }

    private fun dispatch(event: String, payload: String) {
        runCatching {
            when (event) {
                "device-list" -> {
                    val parsed = json.decodeFromString<DeviceListEvent>(payload)
                    _devices.value = parsed.payload.devices
                    parsed.payload.devices.firstOrNull { it.isActive }?.let { _activeDeviceId.value = it.id }
                }
                "state" -> {
                    val parsed = json.decodeFromString<SyncStateEvent>(payload)
                    if (parsed.fromDeviceId != _deviceId.value) {
                        _remoteState.value = parsed.payload
                        parsed.payload.activeDeviceId?.let { _activeDeviceId.value = it }
                    }
                }
                "claim" -> {
                    val parsed = json.decodeFromString<SyncClaimEvent>(payload)
                    val from = parsed.fromDeviceId ?: return@runCatching
                    _activeDeviceId.value = from
                    if (from != _deviceId.value) onRemoteClaim?.invoke(parsed.payload.deviceName)
                }
                "command" -> {
                    val parsed = json.decodeFromString<SyncCommandEvent>(payload)
                    val target = parsed.targetDeviceId
                    val forUs = target == null || target == _deviceId.value
                    // Untargeted commands only apply to whoever holds playback.
                    if (forUs && (target != null || isThisDeviceActive)) {
                        onCommand?.invoke(parsed.payload)
                    }
                }
                "disconnect" -> {
                    val gone = json.decodeFromString<SyncClaimEvent>(payload).fromDeviceId
                    if (gone != null) {
                        _devices.value = _devices.value.filterNot { it.id == gone }
                        if (_activeDeviceId.value == gone) _activeDeviceId.value = null
                    }
                }
            }
        }.onFailure { Log.w(TAG, "bad sync frame ($event): ${it.message}") }
    }

    // -- outbound -----------------------------------------------------------

    /** Announces this device as the one playing audio. */
    fun claim() {
        val id = _deviceId.value.ifBlank { return }
        _activeDeviceId.value = id
        publish(
            buildJsonObject {
                put("type", "claim")
                put("fromDeviceId", id)
                put("payload", buildJsonObject { put("deviceName", deviceName) })
            }
        )
    }

    /** Asks another device to take over playback. */
    fun transferTo(targetDeviceId: String) {
        val id = _deviceId.value.ifBlank { return }
        publish(
            buildJsonObject {
                put("type", "command")
                put("fromDeviceId", id)
                put("targetDeviceId", targetDeviceId)
                put("payload", buildJsonObject { put("action", "claim") })
            }
        )
    }

    /** Sends a transport command to the device currently holding playback. */
    fun sendCommand(action: String, seconds: Double? = null, volume: Double? = null, trackId: String? = null) {
        val id = _deviceId.value.ifBlank { return }
        val target = _activeDeviceId.value
        publish(
            buildJsonObject {
                put("type", "command")
                put("fromDeviceId", id)
                if (target != null && target != id) put("targetDeviceId", target)
                put(
                    "payload",
                    buildJsonObject {
                        put("action", action)
                        seconds?.let { put("seconds", it) }
                        volume?.let { put("volume", it) }
                        trackId?.let { put("trackId", it) }
                    }
                )
            }
        )
    }

    /** Broadcasts what this device is playing so other clients can mirror it. */
    fun publishState(
        currentTrack: Track?,
        isPlaying: Boolean,
        positionSeconds: Double,
        durationSeconds: Double,
        queue: List<Track>,
        currentIndex: Int
    ) {
        val id = _deviceId.value.ifBlank { return }
        if (!repo.settings.value.syncEnabled) return
        val trackJson: JsonElement = currentTrack?.let { json.encodeToJsonElement(it) } ?: JsonObject(emptyMap())
        // Long queues are trimmed: the payload only exists so remote UIs can
        // render "up next", and the server relays it to every device.
        val queueJson = json.encodeToJsonElement(queue.take(50))
        publish(
            buildJsonObject {
                put("type", "state")
                put("fromDeviceId", id)
                put(
                    "payload",
                    buildJsonObject {
                        put("trackId", currentTrack?.id)
                        put("currentTrack", trackJson)
                        put("isPlaying", isPlaying)
                        put("currentTime", positionSeconds)
                        put("duration", durationSeconds)
                        put("queue", queueJson)
                        put("currentIndex", currentIndex)
                        put("activeDeviceId", id)
                    }
                )
            }
        )
    }

    private fun publish(body: JsonObject) {
        scope.launch(Dispatchers.IO) {
            runCatching {
                val config = repo.config.value
                if (!config.isConfigured) return@runCatching
                val request = Request.Builder()
                    .url(ApiClient.normalizedBaseUrl(config) + "/api/sync/publish")
                    .post(json.encodeToString(body).toRequestBody(JSON_MEDIA_TYPE))
                    .build()
                ApiClient.okHttp(config).newCall(request).execute().close()
            }.onFailure { Log.w(TAG, "sync publish failed: ${it.message}") }
        }
    }

    private fun enc(value: String) = java.net.URLEncoder.encode(value, "UTF-8")

    private companion object {
        const val TAG = "MusicySync"
        val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

File: mobile/iOS/Sources/Sync/SyncClient.swift
import Combine
import Foundation
import UIKit

/// Musicy Connect for iOS — joins the same device bus the web app uses, so
/// playback can be handed between a browser tab and the phone.
///
/// The transport is the server's SSE stream read with `URLSession.bytes`;
/// commands go back over `/api/sync/publish`.
@MainActor
final class SyncClient: ObservableObject {
    static let shared = SyncClient()

    @Published private(set) var devices: [SyncDevice] = []
    @Published private(set) var activeDeviceId: String?
    @Published private(set) var isConnected = false

    /// Stable per-install id. Other devices see this in the picker.
    let deviceId: String = {
        let key = "musicy_device_id"
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty { return existing }
        let suffix = UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(16)
        let generated = "dev_ios_" + String(suffix)
        UserDefaults.standard.set(generated, forKey: key)
        return generated
    }()

    var deviceName: String {
        get { UserDefaults.standard.string(forKey: "musicy_device_name") ?? UIDevice.current.name }
        set { UserDefaults.standard.set(newValue, forKey: "musicy_device_name") }
    }

    var isEnabled: Bool {
        get { (UserDefaults.standard.object(forKey: "musicy_sync_enabled") as? Bool) ?? true }
        set {
            UserDefaults.standard.set(newValue, forKey: "musicy_sync_enabled")
            if newValue { start() } else { stop() }
        }
    }

    var isThisDeviceActive: Bool { activeDeviceId != nil && activeDeviceId == deviceId }

    /// True when another device holds playback and we act as a remote.
    var isRemoteControlling: Bool { activeDeviceId != nil && activeDeviceId != deviceId }

    private var streamTask: Task<Void, Never>?
    private var heartbeatTask: Task<Void, Never>?
    private var remoteState: SyncStateEvent.Payload?

    private init() {}

    func start() {
        guard MusicyAPI.shared.isAuthenticated, isEnabled, streamTask == nil else { return }

        AudioPlayer.shared.onStateChanged = { [weak self] in
            Task { @MainActor in self?.publishStateIfActive() }
        }

        streamTask = Task { [weak self] in
            var attempt = 0
            while !Task.isCancelled {
                guard let self else { return }
                do {
                    try await self.openStream()
                    attempt = 0
                } catch {
                    self.isConnected = false
                }
                if Task.isCancelled { return }
                // Exponential backoff with jitter, capped, matching the web client.
                let delay = min(15.0, pow(2.0, Double(min(attempt, 4))))
                attempt += 1
                try? await Task.sleep(nanoseconds: UInt64((delay + Double.random(in: 0...0.5)) * 1_000_000_000))
            }
        }

        heartbeatTask = Task { [weak self] in
            while !Task.isCancelled {
                let playing = self?.isThisDeviceActive == true && AudioPlayer.shared.isPlaying
                try? await Task.sleep(nanoseconds: playing ? 1_000_000_000 : 3_000_000_000)
                self?.publishStateIfActive()
            }
        }

        Task {
            if let list = try? await MusicyAPI.shared.getDevices() {
                devices = list
                if let active = list.first(where: { $0.isActive == true }) { activeDeviceId = active.id }
            }
        }
    }

    func stop() {
        streamTask?.cancel()
        heartbeatTask?.cancel()
        streamTask = nil
        heartbeatTask = nil
        isConnected = false
    }

    private func openStream() async throws {
        let name = deviceName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "iPhone"
        let path = "api/sync/stream?deviceId=\(deviceId)&name=\(name)"
        guard var request = MusicyAPI.shared.makeRequest(path: path) else { throw APIError.invalidURL }
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.timeoutInterval = .infinity

        let (bytes, response) = try await URLSession.shared.bytes(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode < 400 else {
            throw APIError.server((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        isConnected = true

        var event = "message"
        var data = ""
        for try await line in bytes.lines {
            if line.isEmpty {
                if !data.isEmpty { handle(event: event, payload: data) }
                event = "message"
                data = ""
            } else if line.hasPrefix(":") {
                continue // heartbeat comment
            } else if line.hasPrefix("event:") {
                event = String(line.dropFirst(6)).trimmingCharacters(in: .whitespaces)
            } else if line.hasPrefix("data:") {
                data += String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces)
            }
        }
        isConnected = false
    }

    private func handle(event: String, payload: String) {
        guard let data = payload.data(using: .utf8) else { return }
        let decoder = JSONDecoder()

        switch event {
        case "device-list":
            guard let parsed = try? decoder.decode(DeviceListEvent.self, from: data) else { return }
            devices = parsed.payload.devices
            if let active = parsed.payload.devices.first(where: { $0.isActive == true }) {
                activeDeviceId = active.id
            }

        case "state":
            guard let parsed = try? decoder.decode(SyncStateEvent.self, from: data),
                  parsed.fromDeviceId != deviceId else { return }
            remoteState = parsed.payload
            if let active = parsed.payload.activeDeviceId { activeDeviceId = active }

        case "claim":
            guard let parsed = try? decoder.decode(SyncClaimEvent.self, from: data),
                  let from = parsed.fromDeviceId else { return }
            activeDeviceId = from
            // Another device took over: stop our audio immediately.
            if from != deviceId { AudioPlayer.shared.player.pause() }

        case "command":
            guard let parsed = try? decoder.decode(SyncCommandEvent.self, from: data) else { return }
            let target = parsed.targetDeviceId
            let forUs = target == nil || target == deviceId
            // Untargeted commands only apply to whoever holds playback.
            guard forUs, target != nil || isThisDeviceActive else { return }
            apply(parsed.payload)

        case "disconnect":
            guard let parsed = try? decoder.decode(SyncClaimEvent.self, from: data),
                  let gone = parsed.fromDeviceId else { return }
            devices.removeAll { $0.id == gone }
            if activeDeviceId == gone { activeDeviceId = nil }

        default:
            break
        }
    }

    private func apply(_ command: SyncCommandEvent.Payload) {
        let player = AudioPlayer.shared
        switch command.action {
        case "play": player.player.play()
        case "pause": player.player.pause()
        case "toggle": player.toggle()
        case "next": player.next()
        case "previous": player.previous()
        case "seek": if let seconds = command.seconds { player.seek(to: seconds) }
        case "setVolume": if let volume = command.volume { player.setVolume(Float(volume)) }
        case "claim":
            claim()
            applyRemoteState()
        case "playTrack":
            if let queued = command.queue, !queued.isEmpty {
                let start = command.currentIndex
                    ?? queued.firstIndex(where: { $0.id == command.trackId })
                    ?? 0
                player.play(tracks: queued, startAt: start)
            } else if let trackId = command.trackId {
                Task {
                    if let track = try? await MusicyAPI.shared.getTrack(id: trackId) {
                        player.play(tracks: [track])
                    }
                }
            }
        case "shuffle":
            player.toggleShuffle()
        case "setRepeat":
            switch command.mode {
            case "one", "track": player.repeatMode = .one
            case "all", "playlist": player.repeatMode = .all
            default: player.repeatMode = .off
            }
        default:
            break
        }
    }

    // MARK: - Outbound

    private func applyRemoteState() {
        let player = AudioPlayer.shared
        guard let state = remoteState else {
            player.player.play()
            return
        }
        if let queue = state.queue, !queue.isEmpty {
            let start = state.currentIndex
                ?? queue.firstIndex(where: { $0.id == state.trackId })
                ?? 0
            player.play(tracks: queue, startAt: start)
            if let time = state.currentTime, time > 1 { player.seek(to: time) }
        } else if let track = state.currentTrack {
            player.play(tracks: [track])
            if let time = state.currentTime, time > 1 { player.seek(to: time) }
        }
        if let shuffle = state.shuffle { player.shuffle = shuffle }
        switch state.repeatMode {
        case "one", "track": player.repeatMode = .one
        case "all", "playlist": player.repeatMode = .all
        case "off": player.repeatMode = .off
        default: break
        }
        if state.isPlaying == false { player.player.pause() }
    }

    /// Announces this device as the one playing audio.
    func claim() {
        activeDeviceId = deviceId
        publish([
            "type": "claim",
            "fromDeviceId": deviceId,
            "payload": ["deviceName": deviceName]
        ])
    }

    /// Asks another device to take over playback.
    func transfer(to device: SyncDevice) {
        publishStateIfActive()
        publish([
            "type": "command",
            "fromDeviceId": deviceId,
            "targetDeviceId": device.id,
            "payload": ["action": "claim"]
        ])
    }

    /// Sends a transport command to whichever device is currently playing.
    func sendCommand(_ action: String, seconds: Double? = nil, mode: String? = nil) {
        var payload: [String: Any] = ["action": action]
        if let seconds { payload["seconds"] = seconds }
        if let mode { payload["mode"] = mode }
        var body: [String: Any] = [
            "type": "command",
            "fromDeviceId": deviceId,
            "payload": payload
        ]
        if let active = activeDeviceId, active != deviceId { body["targetDeviceId"] = active }
        publish(body)
    }

    private func publishStateIfActive() {
        guard isEnabled, isThisDeviceActive else { return }
        let player = AudioPlayer.shared
        guard let track = player.currentTrack,
              let trackData = try? JSONEncoder().encode(track),
              let trackJSON = try? JSONSerialization.jsonObject(with: trackData) else { return }

        let queueJSON = (try? JSONEncoder().encode(Array(player.queue.prefix(80)))).flatMap {
            try? JSONSerialization.jsonObject(with: $0)
        } ?? []
        let repeatMode: String = {
            switch player.repeatMode {
            case .one: return "track"
            case .all: return "playlist"
            case .off: return "off"
            }
        }()
        publish([
            "type": "state",
            "fromDeviceId": deviceId,
            "payload": [
                "trackId": track.id,
                "currentTrack": trackJSON,
                "isPlaying": player.isPlaying,
                "currentTime": player.position,
                "duration": player.duration,
                "queue": queueJSON,
                "currentIndex": player.currentIndex,
                "shuffle": player.shuffle,
                "repeatMode": repeatMode,
                "activeDeviceId": deviceId
            ] as [String: Any]
        ])
    }

    private func publish(_ body: [String: Any]) {
        Task { await MusicyAPI.shared.publishSync(body) }
    }
}
